src/NServiceBus.Core.Tests/Pipeline/TestableMessageOperations.cs

```diff
@@ -0,0 +1,35 @@
+﻿namespace NServiceBus.Core.Tests.Pipeline
+{
+    using System;
+    using NServiceBus.Pipeline;
+    using System.Threading.Tasks;
+    using MessageInterfaces.MessageMapper.Reflection;
+
+    class TestableMessageOperations : MessageOperations
+    {
+        public Pipeline<IOutgoingPublishContext> PublishPipeline => (Pipeline<IOutgoingPublishContext>)publishPipeline;
+        public Pipeline<IOutgoingSendContext> SendPipeline => (Pipeline<IOutgoingSendContext>)sendPipeline;
+        public Pipeline<IOutgoingReplyContext> ReplyPipeline => (Pipeline<IOutgoingReplyContext>)replyPipeline;
+        public Pipeline<ISubscribeContext> SubscribePipeline => (Pipeline<ISubscribeContext>)subscribePipeline;
+        public Pipeline<IUnsubscribeContext> UnsubscribePipeline => (Pipeline<IUnsubscribeContext>)unsubscribePipeline;
+
+        public TestableMessageOperations() : base(new MessageMapper(), new Pipeline<IOutgoingPublishContext>(), new Pipeline<IOutgoingSendContext>(), new Pipeline<IOutgoingReplyContext>(), new Pipeline<ISubscribeContext>(), new Pipeline<IUnsubscribeContext>(), new ActivityFactory())
+        {
+        }
+
+        public class Pipeline<TContext> : IPipeline<TContext> where TContext : IBehaviorContext
+        {
+            public Action<TContext> OnInvoke { get; set; }
+
+            public TContext Context { get; private set; }
+
+            public Task Invoke(TContext context)
+            {
+                Context = context;
+                OnInvoke?.Invoke(context);
```

Comment:
> By assigning a default delegate you could remove the null check.

that would require a delegate to be assigned to each instance of the pipeline though. Not sure I see a huge benefit over the current approach?

> From a general perspective do we tend to make those delegates not return Task because we would return a completed task anyway most of the time? Do we have other occurrences of this to crosscheck?

saving all the `return Task.CompletedTask` lines was the main goal of not making this async because I don't think there are any known cases where we need to call async APIs in the tests as part of the pipeline. This can easily be changed if needed (or solved via a custom implementation if it's a one-off) imo